File: Queue/Sources/Queue/Queue.swift
import Foundation

public struct Queue<T> {

	private var data: [T?] = []
	private var head: Int = 0

	public var isEmpty: Bool {
		return self.count == 0
	}

	public var count: Int {
		return self.data.count - self.head
	}

	public var front: T? {
		return (self.isEmpty) ? nil : self.data[self.head]
	}

	public mutating func enqueue(_ item: T) {
		self.data.append(item);
	}

	public mutating func dequeue() -> T? {
		guard self.head < self.data.count, let item = self.data[head] else {
			return nil
		}

		self.data[head] = nil
		self.head = self.head + 1

		let totalNumberOfUsedSlots = self.data.count
		let percentageOfEmptySlots = Double(self.head) / Double(totalNumberOfUsedSlots)

		if totalNumberOfUsedSlots > 50 && percentageOfEmptySlots > 0.25 {
			self.data.removeFirst(self.head)
			self.head = 0
		}

		return item
	}

}

File: Queue/Javascript/src/Queue.js

class Queue {

	constructor() {
		this._elements = [];
		this._size = 0;
	}

	get isEmpty() {
		return (this._size == 0);
	}

	get size() {
		return this._size;
	}

	peek() {
		if (this._size == 0) return null;
		return this._elements[0];
	}

	enqueue(element) {
		this._elements.push(element);
		this._size += 1;
	}

	dequeue() {
		if (this._size == 0) return null;
		this._size -= 1;
		return this._elements.shift();
	}

}
module.exports = Queue;
